src/location.rs

```diff
@@ -6,77 +6,86 @@
 // KIND, either express or implied. Please review the Licences for the specific language governing
 // permissions and limitations relating to use of the SAFE Network Software.
 
-use crate::xor_space::{Prefix, XorName};
-use std::fmt::{self, Debug, Formatter};
+use crate::{
+    error::{Result, RoutingError},
+    id::PublicId,
+    xor_space::{Prefix, XorName},
+};
 
-/// A `Location` is a source or destination of a message.
-///
-/// `Node` is single-node location (i.e. no verification of messages from
-/// additional sources needed). It's name is the `Authority::key` other
-/// locations require agreement by a quorum of `Elders`.
-#[derive(Serialize, Deserialize, PartialEq, PartialOrd, Eq, Ord, Clone, Copy, Hash)]
-pub enum Location {
-    /// A single section whose prefix matches the given name
-    Section(XorName),
-    /// A set of nodes with names sharing a common prefix - may span multiple `Section`s present in
-    /// the routing table or only a part of a `Section`
-    PrefixSection(Prefix<XorName>),
-    /// A single node
-    Node(XorName),
+/// Message source location.
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+pub enum SrcLocation {
+    /// A single node with the given public id.
+    Node(PublicId),
+    /// A section with the given prefix.
+    Section(Prefix<XorName>),
 }
 
-impl Location {
-    /// Returns `true` if the location consists of multiple nodes, otherwise `false`.
-    pub fn is_multiple(&self) -> bool {
+impl SrcLocation {
+    /// Returns `true` if the location is a single node, and `false` otherwise.
+    pub fn is_single(&self) -> bool {
         match self {
-            Self::Section(_) | Self::PrefixSection(_) => true,
-            Self::Node(_) => false,
+            Self::Node(_) => true,
+            Self::Section(_) => false,
         }
     }
 
+    /// Returns `true` if the location consists of multiple nodes, otherwise `false`.
+    pub fn is_multiple(&self) -> bool {
+        !self.is_single()
+    }
+}
+
+/// Message destination location.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+pub enum DstLocation {
+    /// Destination is a single node with the given name.
+    Node(XorName),
+    /// Destination are the nodes of the section whose prefix matches the given name.
+    Section(XorName),
+    /// Destination are the nodes whose name is matched by the given prefix.
+    Prefix(Prefix<XorName>),
+    /// Destination is the node at the `ConnectionInfo` the message is directly sent to.
+    Direct,
```

Comment:
I think this should go as well. We will always have the connectioninfo to send the message to. So when it's a message to forward the receiver only needs to check dst == him or forward, but we cannot send a message to anyone we don't have the connection info for. So here we use Direct, but I think that itsef may be soemthing the `send` method gives to `quic-p2p` when it selects a target Feels like we perhaps mix this up a little here. Maybe the unification push I have atm but the actual wire format will need to include the `<SocketAddr>` of the target and we can ignore the Xorname there. So the routing message has no direct, but the `quic-p2p` message needs a SocketAddr and the message. Maybe that is the answer?

Reply:
Could be a follow up PR if that make sense, maybe?

Reply:
So `DstLocation::Direct` does several things:
1. It's for messages sent directly to a socket address we know, so no hops/relays take place
2. We don't know or don't care about the recipient's name
3. The destination is whoever actually receives the message (so `in_location(Direct)` is always `true`)

I added this variant to keep the "direct message" flows we currently have in order to keep the amount of changes in this PR somehow reasonable. I don't mind trying to addres this in a followup PR though.